### PR TITLE
Include json formatter

### PR DESCRIPTION
**Pull request by fede-moya**
As discussed in [https://github.com/codeclimate/test-reporter/issues/413](https://github.com/codeclimate/test-reporter/issues/413)  , this PR has the intention of both adding the capability to modify the formatter through an environmental variable and also to include a json formatter by default.

We are suggesting to use the following json formatter https://github.com/codeclimate-community/simplecov_json_formatter. 

Additionally, as part of the original discussion, it would be the extreme help for those developers using CodeClimate services if we add the possibility to output JSON when the `CC_TEST_REPORTER_ID` is present. 

---
**TO-DO**

- [x] Get agreement on the implementation
- [x] Add unit tests
- [x] Add cucumber/feature tests
- [x] Add documentation to express this new functionality

### CODE REVIEW

**Comment by fede-moya** [comment]
@PragTob Hello Tobias, how are you ? What do you think of the implementation ?

**Comment by PragTob** [comment]
:wave: 

Hi there, sorry I'm not getting much open source work in these days during the week, I'll hopefully take a look on Saturday. Sorry!

Tobi

**Review by PragTob — COMMENTED** [review]
Hi, thanks for the PR :tada: 

See the inline comments, but most of them are just about how the goal should be accomplished and that I'd rather keep the environment variable magic out of there for now :)

also sorry for the CI, I've got no idea what's wrong with it.

The more complicated matter is the question of where the code should live. I think we might have had a misunderstanding over at the test-reporter issue:

I want to move this to a mono repo with `simplecov` and `simplecov-html` and potentially `simplecov-core` or something like that. The problem I've realized with having the repos separate is that they're usually so tightly integrated that they break all too easy with unintended changes (especially in core). 

With HTML it's even a bit worse since its tests live here, but ultimately the JSON acceptance/integration tests would probably also live here because the setup/helpers for it are already here.

From a maintenance/me work perspecitve I'd rather have the repo with you and me not worrying about it of course, but from the perspective of this shipping as a default functionality of SimpleCov now that might break things I need at least push access (git and gem) (not just me but other SimpleCov maintainers as well).

Still, there's always worry about it breaking. In my experience the easiest way to achieve this is have it in a mono repo were all tests run once anything has changed. Of course I haven't done any mono repo migration yet (and I'm unsure when I'd get to it) so that'sa bit up in the air sadly.

That of course would also mean me reviewing all of the formatter code (as it'd land in her) and having some tests for it in our cucumber/aruba setup.

What do you think?


IMG_20190915_142942

**Comment by PragTob** [comment]
For ref about the CI the issue is t #925

File: doc/alternate-formatters.md [hunk 1]
```diff
@@ -35,8 +35,8 @@ CSV formatter for SimpleCov
 
 Flexible formatter that generates the output from an ERB template.
 
-#### [simplecov-json](https://github.com/vicentllongo/simplecov-json)
-*by Vicent Llongo*
+#### [simplecov-json](https://github.com/codeclimate-community/simplecov_json_formatter)
+*by CodeClimate Community"
```
[the remark this answers is not in the document]
Comment (fede-moya) [reply doc/alternate-formatters.md:39]: ooops, sorry I thought it was currently unmaintained 🙏🏼

File: lib/simplecov/default_formatter.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+DEFAULT_FORMATTER = "HTML"
+
+def constantize_formatter(formatter)
```
[the remark this answers is not in the document]
Comment (fede-moya) [reply lib/simplecov/default_formatter.rb:5]: Makes sense 👍🏼

File: simplecov.gemspec [hunk 3]
```diff
@@ -37,6 +37,7 @@ Gem::Specification.new do |gem|
 
   gem.add_dependency "docile", "~> 1.1"
   gem.add_dependency "simplecov-html", "~> 0.11"
+  gem.add_dependency "simplecov_json_formatter", "~> 0.1.2"
```
[the remark this answers is not in the document]
Comment (fede-moya) [reply simplecov.gemspec:40]: Makes sense 👍🏼

**Comment by fede-moya** [comment]
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

**Comment by fede-moya** [comment]
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

**Comment by fede-moya** [comment]
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

If one of your ideas described back in the test-reporter issue was to perhaps start the monorepo migration with having this JSON Formatter living inside this repository, sorry I might have misunderstood that. Saying that,  when reading the test reporter issue I couldn't thought on an implementation of the `simplecov_json_formatter` that could facilitate your monorepo's migration plan, though I do realize that including the `simplecov_json_formatter` will indeed increase the amount of work needed for making this repo a monorepo. We'll be happy to help you with that migration after you structure it.

**Comment by fede-moya** [comment]
Thanks for taking some weekend time for this Tobias 👏🏻

Regarding the environment variable magic, makes sense to exist on it's own PR. I would also enjoy removing the use of `const_get`. So, following your feedback I extract that from this PR.

If one of your ideas described back in the test-reporter issue was to perhaps start the monorepo migration with having this JSON Formatter living inside this repository, sorry I might have misunderstood that. Saying that,  when reading the test reporter issue I couldn't thought on an implementation of the `simplecov_json_formatter` that could facilitate your monorepo's migration plan, though I do realize that including the `simplecov_json_formatter` will indeed increase the amount of work needed for making this repo a monorepo. We'll be happy to help you with that migration after you structure it.

I will grant you with access for both git and gem. 

Sounds good ?

**Comment by fede-moya** [comment]
I have already added you as an owner of the gem. I think that enables you to add others.

**Comment by fede-moya** [comment]
Feedback addressed. Removed the environment variable magic, simplified `default_formatter.rb` and load `simplecov_json_formatter` on demand.

Tobias, would you like to add a reference in the docs of this new "capabilty", to keep record of it ? I can add it, just I didn't because I wasn't sure If you want it since its not a "general feature" lets say.

**Comment by fede-moya** [comment]
Could unit tests be enough ? or should we write some cucumber tests ?

**Comment by PragTob** [comment]
Hi there @fede-moya ,

sorry for the silence. I've been aiming what little time I had at trying to fix the CI that magically seems to be working again (probably/maybe a new Chrome version that fixed it or they rolled something back).

Yeah having gem and repo access should be good enough for now, thank you :) 

imo it definitely belongs into the docs, at least that there is the JSON formatter now. I think that's very important to mention and good to discover. That it automatically runs for Code Climate I'd mention in its description not as a major thing.

as for tests, for very integrated features such as this I'd prefer to have cucumber/feature tests in the vain of:

* configure it to run with JSON formatter, it runs and the output is correct (we also test the output of the HTML formatter in here, which is part of the reason why that should be a mono repo :D )
* set the code climate environment variable and see that it triggers and gives the right output

File: lib/simplecov/default_formatter.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+module SimpleCov
+  module Formatter
+    class << self
+      def from_env
+        formatters = [SimpleCov::Formatter::HTMLFormatter]
+
+        # When running under a CI that uses CodeClimate, JSON output is expected
+        if ENV.fetch("CC_TEST_REPORTER_ID", nil)
+          require "simplecov_json_formatter"
+          formatters.push(SimpleCov::Formatter::JSONFormatter)
```
Comment (PragTob) [inline lib/simplecov/default_formatter.rb:13]: :+1:

File: spec/default_formatter_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+describe SimpleCov::Formatter do
+  describe ".from_env" do
+    context "when CC_TEST_REPORTER_ID environment variable is set" do
+      before do
+        ENV["CC_TEST_REPORTER_ID"] = "4c9f1de6193f30799e9a5d5c082692abecc1fd2c6aa62c621af7b2a910761970"
+      end
+
+      it "returns a MultiFormatter instance with the HTML and JSON formatter" do
+        expect(SimpleCov::Formatter::MultiFormatter).to receive(:new).with([
```
Comment (PragTob) [inline spec/default_formatter_spec.rb:14]: I think you could make this a lot easier to test by:

1. passing in the ENV as an argument, that way it's easy to test and we don't need to mess with global state that might leak into other tests for weird results (if the tests here are run in random order so that the nil one isn't run some test further down the line might still have it set)
2. just returning the array of formatters to use instead of the Multiformatter and doing that on the call site, that way you avoid the stubbing :)

**Review by PragTob — COMMENTED** [review]
Looking good :+1: 

Small comment about the testing and well feature/acceptance level tests would be great as pointed out in my other comment.

Thanks for implementing all the suggestions so diligently :green_heart: 

WhatsApp Image 2020-06-07 at 08 16 07a

**Comment by fede-moya** [comment]
@PragTob thank you very much for your feedback Tobias and for finding time for this !, I'll be trying to wrap this before the end of the current week.

File: spec/default_formatter_spec.rb (hunk 5 above)
Comment (fede-moya) [reply spec/default_formatter_spec.rb]: Thanks, both suggestions have simplified testing a lot !

**Comment by fede-moya** [comment]
Hey @PragTob 👋🏼  

I have just addressed your latest feedback, added documentation and feature tests. Also I would like to point that you can find more tests for the `JSONFormatter` here https://github.com/codeclimate-community/simplecov_json_formatter/tree/master/spec. 

Let me know If there is anything I should change or add. 

Regards !

**Comment by efueger** [comment]
Hi @PragTob 👋 - Hope all is well! Just a quick check-in to see if you need anything else from this side.

**Comment by PragTob** [comment]
Hi folks,

no all good. I just had a couple of rough weeks+ in which I didn't have any mind space to dedicate to open source so minus the small bugfix release I didn't touch any OSS in that time so nothing happened here. Sorry about that.

I'll try to cut a release today.

@fede-moya I was aware of the extra tests, but esp. with simplecov being so wide spread I really like to have the additional layer of integration tests to make sure we don't break anything for a whole user run setting. SimpleCov breaks in the most amazing and weird ways.

File: features/config_json_formatter.feature [hunk 6]
```diff
@@ -0,0 +1,43 @@
+@test_unit @config
+Feature:
+
+  SimpleCov::Formatter::JSONFormatter is one of the
+  formatters included by default, useful for exporting
+  coverage results in JSON format.
+
+  Background:
+    Given I'm working on the project "faked_project"
+  Scenario: With JSONFormatter
```
Comment (PragTob) [inline features/config_json_formatter.feature:10]: separate (comment for myself when fixing it up)

File: features/config_json_formatter.feature [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      """
+
+    When I successfully run `bundle exec rake test`
+    Then a JSON coverage report should have been generated in "coverage"
+    And the output should contain "JSON Coverage report generated"
+
+  Scenario: When CC_TEST_REPORTER_ID is set in the environment
+    Given SimpleCov for Test/Unit is configured with:
+      """
+      ENV['CC_TEST_REPORTER_ID'] = "9719ac886877886b7e325d1e828373114f633683e429107d1221d25270baeabf"
```
Comment (PragTob) [inline features/config_json_formatter.feature:31]: move into invoking command and make id more generic/show it's fake (comment for myself when fixing it up)

File: features/step_definitions/json_steps.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Then /^the JSON coverage report should map:/ do |expected_report|
+  cd(".") do
```
Comment (PragTob) [inline features/step_definitions/json_steps.rb:4]: Do we need the `cd`? (comment for myself when fixing it up)

File: features/step_definitions/json_steps.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Then /^the JSON coverage report should map:/ do |expected_report|
```
Comment (PragTob) [inline features/step_definitions/json_steps.rb:3]: Thanks for putting stuff in a separate file!

Comment (PragTob) [inline features/step_definitions/json_steps.rb:3]: add `$` to end of rege

File: features/step_definitions/json_steps.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+  cd(".") do
+    json_report = File.open("coverage/coverage.json").read
```
Comment (PragTob) [inline features/step_definitions/json_steps.rb:5]: File.read? (comment for myself when fixing it up)

File: features/step_definitions/json_steps.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+    expected_report = ERB.new(expected_report).result(binding)
+    expect(json_report).to eq(expected_report)
```
Comment (PragTob) [inline features/step_definitions/json_steps.rb:7]: potentially make it check against partials, ERB evaluation seems too involved. (comment for myself when fixing it up)

**Comment by PragTob** [comment]
Was merged via #950

**Comment by PragTob** [comment]
sorry was in a hurry to get this out and felt bad about leaving this here for so long, forgot the customary bunny and thanks :green_heart: 

ghost
